### ISSUE TITLE
Database dumps do not work on large databases

### ISSUE THREAD

**Issue by Brayden** [outerbase/starbasedb#59]
**Describe the bug**
If you try to use any of the database dump endpoints such as SQL, CSV or JSON the data is loaded into memory and then created as a dump file. To support any size database we should likely investigate enhancements such as streaming or file stitching (keeping open-minded) to allow any size of database to be exported. Perhaps options for dumping subsets of data at a time too (e.g. one table, or 10,000 rows, etc).

**To Reproduce**

1. Hit the `/export/dump` endpoint on a large database
2. Will eventually fail

**Expected behavior**
As a user I would expect any and all of the specified data to be dumped out without an error and without partial results.

**Screenshots**

**Desktop (please complete the following information):**

**Smartphone (please complete the following information):**

**Additional context**

**Comment by b4s36t4** [comment]
Hi, @Brayden. instead of using DO alarm API we can use queues if the database is size greater than certain size (say 200 or 500)?

My suggestion is to attach the cloudflare queues to the new private endpoints where the consumer and producer are a single worker which will export the data into R2 for certain time say (20 to 30sec) depends on usage limits, this way we can have the queues with the capability of the last record of the export queue and there's blockage of any request/response.

https://developers.cloudflare.com/durable-objects/api/sql-storage/#databasesize we can depend on this property of DO to trigger the queue or something like.

**Comment by Brayden** [comment]
Hey @b4s36t4!

Instead of using the DO alarm API, I think queues is also an acceptable approach. Just to talk through it a bit more so I fully understand how you're thinking of approaching it. A couple of questions:

- When you mention attaching to new private endpoints, would this be a new Worker altogether? Could it instead be a Starbase plugin perhaps? Here's a blog outlining how they work and some examples are also included in the repo (https://www.outerbase.com/blog/extensibility-with-plugins/)
- When you say we can base it off if the database is size greater than some defined number (e.g. 200 or 500) would we fetch the database size at some interval to see if the difference in size from the last export meets the criteria and auto-run the backup?
- Could we allow for users to define say like "backup daily" or "backup weekly" and trigger at that time?

Thanks for looking into this!

**Comment by Kunal-Darekar** [comment]
/attempt #92 

Implementation Plan for Database Dump Feature:

1. Core Implementation:
   - Create chunked processing system for database dumps
   - Implement progress tracking mechanism
   - Add R2 storage integration for dump files
   - Build callback notification system

2. Key Functions to Implement:
   - startChunkedDumpRoute(): Initialize dump process
   - processDumpChunk(): Handle data chunks
   - getDumpFileRoute(): Retrieve dump files
   - finalizeDump(): Complete dump process

3. Testing Strategy:
   - Unit tests for all core functions
   - Integration tests for the complete flow
   - Edge case handling (empty DBs, large datasets)
   - Error scenarios and recovery

4. Features to Cover:
   - Large database support through chunking
   - Progress monitoring
   - SQL escaping
   - Callback notifications
   - R2 storage integration

5. Timeline:
   - Core implementation: Completed
   - Testing: Completed
   - Documentation: Completed
   - PR ready for review

Status: Implementation complete and ready for review. All tests passing.

![Image](https://github.com/user-attachments/assets/6d254193-53be-47a1-8053-fbe6101533eb)

**Comment by Kunal-Darekar** [comment]
@Brayden  Apologies, I mistakenly deleted my original PR while making some update .I’ve created a new PR #93 and included the necessary documentation, including the wrangler.toml variables and the cURL command/endpoint for triggering. Everything should be clear and ready for testing. Could you please check? Thanks for your patience!

**Comment by ezzeddinp** [comment]
hey @Brayden nice to know you!, are you still having these kind of issues? or it has already solved? let me know!, thanks @Brayden

**Comment by whynice724-cell** [comment]
/attempt #59

Implementation plan (primary):
1. Add an async export job flow for large dumps (`/export/dump`) with a small sync fast-path for tiny DBs.
2. Persist job state/checkpoints in DO storage (current table, last row cursor, bytes written, status) so work can resume safely.
3. Stream chunked SQL/CSV/JSON output to R2 in bounded time slices (under request limits), then continue via DO Alarm until complete.
4. Add job status + artifact endpoints (e.g. job id, progress, downloadable URL) and optional callback webhook on completion/failure.
5. Keep DB responsive by yielding between chunks and enforcing cooperative "breathing" windows when request queues are non-empty.
6. Add tests for resume/idempotency/error handling + docs for required bindings/env vars and curl examples.

I will start with SQL dump first (end-to-end), then extend the same job pipeline to CSV/JSON.

**Comment by yuliuyi717-ux** [comment]
/attempt #59

I can take this bounty.

Implementation plan:
1) Implement a resumable export job model (job id + checkpoint state) so dump generation can continue across request/worker time limits.
2) Stream dump chunks to R2 instead of buffering full data in memory, with chunk boundaries and checkpoint metadata persisted in DO storage.
3) Add cooperative scheduling (short work slices + yield) to reduce long DB lock windows and keep normal traffic responsive.
4) Add job status and artifact retrieval endpoints (progress, completion, error state), plus optional callback notification on completion/failure.
5) Cover SQL/CSV/JSON export paths with integration tests for resume/retry/idempotency behavior on large datasets.

If this direction matches expectations, I will open a focused PR with implementation + docs for required bindings.

**Comment by scrollingreel** [comment]
/attempt #59

Implementation Plan:

Chunked SQL export with R2 streaming — Replace in-memory dump with batched table reads (configurable batch size, default 1000 rows). Each batch is serialized to SQL and appended to an R2 object (dump_YYYYMMDD-HHmmss.sql).

DO Alarm for long-running exports — Track export progress (current table, row offset) in DO storage. If approaching the 30s timeout (~25s threshold), persist state and schedule a DO alarm to resume. On alarm trigger, continue from saved offset.

Breathing intervals — Between batches, check if other requests are queued. If so, yield execution for a configurable interval before resuming, preventing DB lock contention.

Dual delivery path:

Fast path: exports completing in <30s return the file directly in the response
Async path: exports exceeding threshold continue via DO alarms, store result in R2, and POST to a user-provided callback URL on completion
R2 binding requirement — Add R2 bucket binding configuration. Document setup in README.

API changes:

POST /export/dump accepts optional callbackUrl parameter
Returns { status: "processing", exportId: "..." } for async exports
GET /export/status/:exportId to check progress
GET /export/download/:exportId to retrieve completed dump from R2

**Comment by justinleeyang** [comment]
I can take bounty #59. I will share a scoped implementation plan and first PR shortly.

**Comment by justinleeyang** [comment]
Quick follow-up on #59: I am still actively working on this. Please let me know if you want me to narrow scope to a smaller first PR.

**Comment by justinleeyang** [comment]
Quick follow-up on #59: I am still actively working on this. Please let me know if you want me to narrow scope to a smaller first PR.

**Comment by justinleeyang** [comment]
No maintainer response yet on #59, I will switch to another open bounty for now and can resume this issue anytime when feedback arrives.

**Comment by Sidd1211** [comment]
/attempt #59

Implementation plan:
- Replace in-memory dump with chunked export
- Stream dump data in batches to storage
- Prevent 30s worker timeout
- Ensure large databases export successfully

**Comment by shanimauryaa** [comment]
/attempt #59
​Implementation Plan:
I will implement a robust chunked export system as proposed:
​R2 Integration: I will use R2 binding to store the .sql dump file instead of keeping it in memory.
​Chunked Processing: Instead of a single fetch, I will process the DB rows in chunks to stay within Durable Object execution limits.
​DO Alarms: I'll utilize Durable Object alarms to handle operations exceeding the 30s timeout, ensuring the process resumes correctly until completion.
​Callback Notification: I will add a webhook/callback URL feature to notify the user once the R2 upload is finished

**Comment by CuboYe** [comment]
/claim 59

Claiming territory for a scoped fix attempt now. If maintainer preference differs, I can adapt quickly.

**Comment by AbhishekMauryaGEEK** [comment]
/claim #59

This PR fixes the current /export/dump implementation by:

streaming the dump instead of buffering the entire database in memory
fetching rows in chunks to reduce timeout/memory pressure
validating identifiers and parameterizing schema lookup
producing a valid SQL dump format (BEGIN TRANSACTION ... COMMIT)
Focused scope: this intentionally fixes the safety and correctness issues in the existing route without introducing new infrastructure.

**Comment by yutubeboss575-create** [comment]
/attempt #59

Implementation Plan:

Use ZIO Streams to fetch data in chunks from the database to avoid OOM (Out of Memory) issues.

Integrate Cloudflare R2 to store the dump file incrementally.

Implement Durable Object Alarms to handle the 30-second execution limit by persisting the current state and resuming the dump process in a new execution context.

Add a callback mechanism to notify the user upon completion.

**Comment by yutubeboss575-create** [comment]
I've tested the build and it's working fine. Ready for review!

**Comment by shanimaury89-art** [comment]
Hi 👋,

I’d like to work on this issue.

From what I understand, the problem occurs because large database dumps are being fully loaded into memory, which causes timeouts and failures for bigger datasets.

I’m planning to explore a chunk-based or streaming approach to handle large exports more efficiently and avoid memory/time limits.

Please let me know if this direction aligns with your expectations or if there are any specific constraints I should be aware of.

Thanks!

**Comment by chengyixu** [comment]
/attempt #59

## Implementation Plan

Replace the in-memory string dump with a **ReadableStream** + **paginated batch reads** to support databases of any size.

**Key changes:**
- `dumpDatabaseRoute` returns a streaming `Response` (ReadableStream) — bytes flush to the client as they are produced
- Rows are fetched in batches of 1000 (LIMIT/OFFSET) so at most one batch is in memory at a time
- `breathe()` (setTimeout 0) yields to the event loop between batches and between tables, so other DO requests can be serviced during a long export
- Parameterised queries + `quoteIdentifier()` for safe handling of table names with special characters
- Dump wrapped in `BEGIN TRANSACTION; … COMMIT;`
- NULL and boolean values rendered correctly as SQL literals

All 8 unit tests pass.

**Comment by chengyixu** [comment]
/attempt #59

**Implementation plan** for chunked async database dump with R2 storage:

1. **Existing `GET /export/dump`** — kept unchanged for small databases
2. **New `POST /export/dump/async`** — starts a background dump job, returns job ID immediately
3. **New `GET /export/dump/status/:jobId`** — poll job state; returns signed R2 download URL on completion
4. **Chunked row fetching** — generator that fetches 1,000 rows per batch so memory stays bounded regardless of table size
5. **R2 multipart upload** — each 5 MiB buffer is uploaded as a separate part (never holds full dump in memory)
6. **Job state persistence** — `tmp_dump_jobs` table in DO SQLite storage, survives worker restarts
7. **Callback URL support** — POST to caller-provided URL when dump completes
8. **Optional R2 binding** — add `[[r2_buckets]] binding = "DUMP_BUCKET"` to wrangler.toml; returns 501 with setup instructions if missing

PR incoming.

**Comment by CuboYe** [comment]
/claim 59

Claiming this bounty issue. I reviewed the constraints (30s limit, large export streaming, resumability) and will open a PR with a staged export approach and async delivery path.

**Comment by juandirahelia-a11y** [comment]
​/attempt #59

**Comment by TeapoyY** [comment]
I've submitted a fix for this issue. The PR implements R2-backed async processing for large database dumps, with chunked table processing to avoid memory exhaustion. See: https://github.com/outerbase/starbasedb/pull/125

**Comment by juandirahelia-a11y** [comment]
Acknowledged. The implementation of asynchronous R2 processing in PR #125 for large database dumps is a solid move to prevent memory exhaustion. We'll be monitoring the scalability closely. Good work

**Comment by bcornish1797** [comment]
Hi, just a note that I submitted PR #107 on March 23rd addressing this issue. It's still open and mergeable. Would appreciate a review when you get a chance. Thanks!

**Comment by juandirahelia-a11y** [comment]
After reviewing the options, the R2-backed asynchronous approach in PR #125 seems to be the most scalable solution for Issue #59. It addresses the memory exhaustion directly. Requesting a final review from the maintainers to get this merged and deployed. Stability is key.

**Comment by armorbreak001** [comment]
/attempt #59

<div id="algora-attempt" />

## Implementation

I've implemented a streaming solution for large database dumps. The key changes:

1. **ReadableStream output** — memory stays O(batch_size) instead of O(total DB size)
2. **Batched row fetching** — LIMIT/OFFSET with 1000 rows per batch
3. **Cooperative multitasking** — yields control between batches so the Durable Object remains responsive
4. **Proper SQL escaping** — handles NULL, bigint, undefined (original only handled strings)

### PR: https://github.com/outerbase/starbasedb/pull/127

All 23 existing export tests pass ✅

**Comment by shehryarakif** [comment]
/attempt #59
Architectural Blueprint for the Async Export
Start an Export: A user calls the /export/dump endpoint, which triggers the process and immediately returns a unique jobId. Your export is now being handled in the background.

Core Processing Engine: The export job is managed by a dedicated Durable Object (DO) for that specific user. The DO holds the state of the export (e.g., its status and the last processed database row) in its persistent storage, ensuring it can survive restarts and timeouts.

Batched, Streaming Data Pipeline: The DO reads data from the database in manageable chunks. It formats each chunk as SQL and streams it directly to an R2 bucket. For large files, using R2's multipart upload API is recommended, as it allows you to upload parts in parallel and resume failed uploads without starting over.

Time-Sliced Execution with Alarms: The DO performs a batch of work (e.g., reading and uploading 1,000 rows) for a set time, like 5 seconds. Before the time is up, it saves its current position to its state, sets a DO Alarm for a few seconds in the future, and then yields. When the alarm fires, the DO wakes up and continues from where it left off.

Completion and Notification: Once all data is processed, the DO finalizes the multipart upload (if used), updates the job status to "completed", and sends a POST request to a user-provided callback URL (a webhook) with the R2 file details. For very large dumps, you can generate a time-limited, presigned URL for the user to download the file directly.

🔧 Implementation Details with Code Examples
1. Managing State with Durable Objects
The Durable Object is the perfect home for the job's state, as it provides strong consistency and survives across alarms.

typescript
// In your Durable Object class (e.g., ExportManagerDO)
export class ExportManagerDO extends DurableObject {
  // --- State stored in DO's persistent storage ---
  async getState(): Promise<ExportJobState> {
    return (await this.ctx.storage.get("state")) || { 
      status: "idle", 
      processedRows: 0, 
      // ... other metadata like R2 key, uploadId, etc.
    };
  }

  async saveState(state: ExportJobState): Promise<void> {
    await this.ctx.storage.put("state", state);
  }
  // ------------------------------------------------
}
2. Streaming the Dump to R2
For exporting large datasets, buffering the entire file in memory is not an option. Instead, you should format and upload data in a continuous stream. The code below provides a conceptual example of how to create a ReadableStream that fetches data, formats it into SQL INSERT statements, and pipes it to R2.

typescript
async function streamDatabaseToR2(env: Env, jobState: ExportJobState): Promise<void> {
  // Assume we have a function to get the total row count and fetch data in batches
  const totalRows = await getTotalRowCount();
  
  // Create a ReadableStream that will be the source of our file
  const readableStream = new ReadableStream({
    async start(controller) {
      // Write the SQL header (e.g., BEGIN TRANSACTION;)
      controller.enqueue(new TextEncoder().encode("BEGIN;\n"));

      let rowsProcessed = jobState.processedRows; // Resume from last saved position
      const BATCH_SIZE = 1000;
      
      while (rowsProcessed < totalRows) {
        // Fetch the next batch of rows from the database
        const rows = await fetchDataBatch(rowsProcessed, BATCH_SIZE);
        for (const row of rows) {
          // Format each row into an INSERT statement
          const sqlLine = `INSERT INTO table_name (col1, col2) VALUES ('${row.col1}', '${row.col2}');\n`;
          // Push the chunk into the stream
          controller.enqueue(new TextEncoder().encode(sqlLine));
        }
        rowsProcessed += rows.length;

        // After processing a batch, check if we're nearing the time limit (e.g., 5 seconds)
        if (isApproachingTimeLimit()) {
          // Save our progress so we can resume later
          jobState.processedRows = rowsProcessed;
          await saveJobStateToDO(jobState);
          
          // Set an alarm to wake up the DO in a few seconds to continue
          await this.ctx.storage.setAlarm(Date.now() + 5000);
          
          // Close the stream for now. The DO will be woken by the alarm.
          controller.close();
          return; 
        }
      }

      // Write the SQL footer (e.g., COMMIT;)
      controller.enqueue(new TextEncoder().encode("COMMIT;\n"));
      controller.close();
      
      // Finalize the job state after the stream is complete
      jobState.status = "completed";
      await saveJobStateToDO(jobState);
      await notifyUserWebhook(env.WEBHOOK_URL, jobState); // Notify user
    }
  });

  // Stream the ReadableStream to R2
  const r2Object = await env.MY_R2_BUCKET.put(`exports/dump_${Date.now()}.sql`, readableStream);
}
3. Scheduling Continuation with Alarms
The Durable Object Alarms API is the key to gracefully handling the 30-second execution limit. The alarm() handler is where you put the logic to resume the export.

typescript
export class ExportManagerDO extends DurableObject {
  // ... state methods ...

  // This is called automatically when the alarm fires
  async alarm() {
    const jobState = await this.getState();
    if (jobState.status === "processing") {
      // Call the function that continues the streaming process.
      // It will pick up from `jobState.processedRows`.
      await streamDatabaseToR2(this.env, jobState); 
    }
  }

  // The HTTP endpoint that kicks off the export
  async fetch(request: Request) {
    const url = new URL(request.url);
    if (url.pathname === "/export/dump") {
      const jobState = await this.getState();
      if (jobState.status === "idle") {
        jobState.status = "processing";
        await this.saveState(jobState);
        
        // Start the process
        await streamDatabaseToR2(this.env, jobState);
        
        return new Response(JSON.stringify({ jobId: this.ctx.id.toString() }), { status: 202 });
      }
      // ... handle other statuses ...
    }
    // ... other routes ...
  }
}
4. Delivering the Dump via Webhooks
Instead of making the user poll an endpoint, you can notify them when the export is finished by POSTing a JSON payload to a user-provided callback URL.

typescript
// In ExportManagerDO
async function notifyUserWebhook(webhookUrl: string, jobState: ExportJobState): Promise<void> {
  // 1. Generate a presigned URL for the R2 object (e.g., valid for 7 days)
  const presignedUrl = await generateR2PresignedUrl(jobState.r2Key);
  
  // 2. Prepare the notification payload
  const payload = {
    event: "export.completed",
    jobId: this.ctx.id.toString(),
    status: "success",
    downloadUrl: presignedUrl,
    expiresAt: new Date(Date.now() + 7 * 24 * 60 * 60 * 1000).toISOString(), // 7 days
    metadata: {
      fileName: jobState.r2Key,
      rowCount: jobState.processedRows,
      fileSize: jobState.fileSize // Might need to fetch from R2 HEAD request
    }
  };

  // 3. POST the payload to the user's webhook
  try {
    const response = await fetch(webhookUrl, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(payload)
    });
    
    if (!response.ok) {
      // Store the failure in DO state for potential manual retry
      jobState.webhookDeliveryStatus = 'failed';
      jobState.lastWebhookAttempt = Date.now();
      await this.saveState(jobState);
    }
  } catch (error) {
    // Network errors, DNS failures, etc.
    console.error(`Webhook delivery failed: ${error}`);
  }
}
⚖️ Trade-Offs and Alternatives
This architecture, while robust, involves some complexity. Here are a few points to consider:

Alternative: Cloudflare Workflows: Cloudflare Workflows is a service built specifically for orchestrating multi-step, long-running tasks like this export process. It could simplify the logic for retries and state management, potentially reducing the amount of custom code you need to write.

Complexity vs. Simplicity: For databases under a certain size (e.g., under 100MB), a simpler, synchronous export might still be acceptable. You could implement a hybrid approach where the system checks the database size first and uses the async method only if it's large.

Cost: While Durable Object and R2 costs are generally low, a high volume of exports could lead to meaningful charges. Be sure to model your expected usage against Cloudflare's pricing to avoid surprises.

/claim #59

**Comment by xyaz1313** [comment]
/attempt #59

I've submitted a PR that fixes this issue: https://github.com/outerbase/starbasedb/pull/129

**Solution**: Rewrote the dump endpoint to use ReadableStream with LIMIT/OFFSET batching instead of loading everything into memory.

Key changes:
- Streams data in batches of 500 rows
- Uses ReadableStream for incremental response delivery  
- 10ms breathing intervals between batches to avoid DB lock contention
- Peak memory reduced from O(database_size) to O(500_rows)

All 25 existing export tests pass.

**Comment by suletetes** [comment]
/attempt #59


### The Solution (Two-Tier Approach)

We’re splitting exports into two paths:

**Tier 1 – Fast & Synchronous** (for small databases)  
If the export can finish comfortably inside the 30-second window, we just do what we do today return the file directly. Zero changes for users with small data.

**Tier 2 – Asynchronous Chunked Export** (for everything else)  
For larger databases we switch to a background job that streams the data in small chunks, saves it to R2, and keeps going until it’s done. The user gets an immediate response with a job ID they can check later.

### How the Async Flow Works

```mermaid
graph TD
    A["Client POST /export/dump"] --> B{"Estimate size"}
    B -->|"Small enough"| C["Return file directly (Tier 1)"]
    B -->|"Too big"| D["Create export job in tmp_export_jobs"]
    D --> E["Return jobId + statusUrl immediately (202 Accepted)"]
    E --> F["Durable Object Alarm Loop"]
    F --> G["Read next batch of rows"]
    G --> H["Format chunk (SQL/JSON/CSV)"]
    H --> I["Append to R2 multipart upload"]
    I --> J["Save progress (last_table, last_rowid, bytes_written)"]
    J --> K{"Still more to do?"}
    K -->|"Yes"| L["Set next alarm in 100ms"]
    L --> F
    K -->|"No"| M["Finish multipart upload + send webhook"]
```

### Why This Design Makes Sense

**1. How do we keep going past 30 seconds?**  
We use Durable Object **alarms** to schedule the next chunk of work. Each alarm:
- Only runs for about 4–5 seconds
- Saves its exact position (which table, which row) in the database
- Schedules the next alarm with a tiny 100 ms pause

That tiny pause gives the Durable Object a chance to handle other requests (WebSockets, RPCs, etc.) so the whole system stays responsive.

**2. Where do we store the export while it’s being built?**  
**Cloudflare R2** it’s the perfect fit because:
- No extra credentials needed (native binding)
- Supports multipart uploads (we can keep appending chunks)
- Zero egress fees
- S3-compatible if the user wants to download it elsewhere

**3. How does the user actually get the file?**  
Three easy options:
- **Poll** – `GET /export/jobs/:jobId` to check progress
- **Webhook** – provide a `callbackUrl` when starting the job; we POST the final download link when done
- **Direct download** – `GET /export/jobs/:jobId/download` gives them a clean redirect to the R2 file

### What We’ll Need to Add

**R2 Bucket** (add to `wrangler.toml`):

```toml
[[r2_buckets]]
binding = "EXPORT_BUCKET"
bucket_name = "starbasedb-exports"
```

**New internal table** to track jobs:

```sql
CREATE TABLE IF NOT EXISTS tmp_export_jobs (
    id TEXT PRIMARY KEY,
    format TEXT NOT NULL,    -- 'sql' | 'json' | 'csv'
    status TEXT NOT NULL, -- 'pending' |'in_progress' | 'completed' | 'failed'
    target_table TEXT,    -- NULL = full dump
    r2_key TEXT NOT NULL,
    r2_upload_id TEXT,
    current_table TEXT,
    current_offset INTEGER DEFAULT 0,
    total_tables INTEGER,
    completed_tables INTEGER DEFAULT 0,
    bytes_written INTEGER DEFAULT 0,
    callback_url TEXT,
    error_message TEXT,
    created_at TEXT DEFAULT (datetime('now')),
    completed_at TEXT
);
```

### New API (super simple)

**Start an export**

```http
POST /export/dump
Content-Type: application/json

{
  "async": true,                    // optional – forces async mode
  "callbackUrl": "https://...",     // optional webhook
  "format": "sql"                   // sql | json | csv
}
```

Response (202 Accepted):

```json
{
  "result": {
    "jobId": "export_20240101-170000_abc123",
    "status": "pending",
    "statusUrl": "/export/jobs/export_20240101-170000_abc123",
    "estimatedTables": 15
  }
}
```

**Check progress**

`GET /export/jobs/:jobId`

**Download the finished file**

`GET /export/jobs/:jobId/download` → redirects to R2 (or streams directly)

### Smart “Breathing” Strategy

Every alarm cycle does this:
1. Start a timer
2. Process ~5000 rows
3. Check how much time has passed
4. If we’re close to 4–5 seconds → save progress and schedule the next alarm in 100 ms
5. If we finish early → keep going

This keeps the system snappy while still making fast progress on big exports.

### File Naming in R2

```
exports/dump_YYYYMMDD-HHmmss.sql
exports/tablename_YYYYMMDD-HHmmss.json
exports/tablename_YYYYMMDD-HHmmss.csv
```

### Error Handling & Recovery

- If an alarm fails, we keep the last known progress and retry after 60 seconds.
- Jobs stuck “in progress” for >10 minutes get marked failed.
- Partial R2 uploads are kept so users can retry or clean up manually.
- We’ll add a cleanup job (cron) later to delete abandoned uploads.

### Backward Compatibility

Everything that works today still works exactly the same:
- Old `GET /export/dump` calls stay synchronous for small databases.
- Async mode is **opt-in** (via the new POST + `"async": true`) or automatically triggered when we detect the export will be too big.

**Comment by Aurenv1** [comment]
/attempt #59
My Plan:
I’m going to fix the memory exhaustion issue by moving away from loading the entire database at once. My approach will involve:

Chunked Streaming: I'll modify the export logic to process the data in continuous chunks, appending them directly to an R2 bucket instead of holding them in memory.

Timeout Handling: To beat the 30-second execution limit, I’ll use Durable Object alarms to checkpoint the progress. If it's about to time out, the alarm will trigger a restart right where it left off.

Completion Notification: I'll add a callback URL so users get notified immediately once the full dump is ready in R2.

**Comment by aadityaranjan01** [comment]
/attempt #59

Implementing a streaming-based database dump to resolve memory and timeout issues on large datasets.

* Replaces in-memory dump with ReadableStream
* Uses LIMIT/OFFSET with ORDER BY rowid for stable pagination
* Avoids full-table executeOperation calls
* Adds breathing intervals to prevent Durable Object blocking

Scoped to a single file (`dump.ts`) with no changes to shared utilities or other export routes.

Opening PR now.

**Comment by akshatd1007** [comment]
Is it still open? I am intrigued.

**Comment by onelpawarai-X** [comment]
/attempt #59

Implementation Strategy (Advanced & Optimized):

To ensure 100% accuracy and handle databases of any size (up to 10GB+), I will implement the following high-performance architecture directly on Cloudflare's infrastructure:

Direct R2 Streaming with Multipart Upload: Instead of loading data into memory, I will use a Multipart Upload stream to append SQL/CSV chunks directly to an R2 bucket. This bypasses the 1GB memory limit of Durable Objects.

State-Persistent Chunking: I will implement a limit/offset mechanism that marks the progress in the DO's internal memory. This ensures that if a timeout occurs, the process knows exactly where to resume.

Adaptive "Breathing Intervals": To prevent database locking and avoid the 30-second execution limit, I will implement asynchronous pauses (interval of 5 seconds on/off). This allows other requests in the queue to process without interference.

Durable Object Alarms for Persistence: I will utilize DO Alarms to re-trigger the export process automatically if it exceeds the execution window.

Completion Callback System: Once the 100% accurate dump is stored in R2, the system will trigger a Webhook/Callback URL to notify the user immediately.

I am executing this as a Crime Stopper Master project, focusing on zero-error execution and maximum efficiency.

**Comment by onelpawarai-X** [comment]
Hello @algora-io, I have implemented the enterprise-grade SQL engine with advanced PRAGMA optimizations and fail-safe streaming to resolve this issue. This contribution ensures infinite scalability and crash prevention for large database dumps.

Bounty Claim: #59
Fixes: #59

**Comment by onelpawarai-X** [comment]
Hello @algora-io,

I have completed the task for Issue #59 and submitted my solution in this PR. I have implemented a scalable SQLite architecture with fail-safe streaming and optimized memory management as requested.

Please track this PR for the bounty claim associated with Issue #59. I am ready for the review and look forward to the approval.

Bounty Details:

Issue Link: #59

Claimant: @sayidilxs-web

Status: Work Complete / Ready for Review

CC: @outerbase

**Comment by Mira-Mjodheim** [comment]
/attempt #59

**Plan:**
- Perform deep trace of the reported issue in the local environment.
- Identify the root cause in the core logic.
- Implement a robust fix with regression tests.

**Comment by brone1323** [comment]
/attempt #59

Plan:

The three export endpoints (`/export/dump`, `/export/json/:tableName`, `/export/csv/:tableName`) all share the same root cause: they call `executeOperation` to load the entire result set into a JS array, build the full payload as a single string, then wrap it in a `Blob` before responding. With a 1-10 GB SQLite DB this OOMs the DO and blocks the event loop long enough for Cloudflare to evict it.

I'm refactoring to:

1. **`ReadableStream` responses**, so bytes flow to the client as soon as they're encoded. No intermediate `Blob` or full-document string.
2. **Keyset/LIMIT-OFFSET paged reads** against each table (default 1000 rows/page) using the existing `executeOperation` interface, so we never materialise more than one page in memory at a time.
3. **Cooperative yields ("breathing intervals")** between pages via `await scheduler.wait(0)` (with a `setTimeout(0)` fallback) so the DO runtime can service alarms, websockets, and storage I/O instead of being killed for hogging the isolate.
4. **Format-specific incremental encoding**:
   - SQL dump: schema first, then `INSERT INTO t VALUES (...);` lines flushed per page.
   - CSV: header row first, then row chunks; preserves the existing quoting rules.
   - JSON: streamed as `[` + comma-separated rows + `]` so the output is still a valid JSON array, no per-row JSON.stringify of the whole set.
5. **Existing tests stay green** (response shape, headers, escaping); add streaming-specific tests that assert paged reads happen and that the body is consumed via the stream rather than buffered.

Will not touch import paths, auth, or the `executeOperation` contract — this is a discrete refactor inside `src/export/`.

Note up front: I cannot record a demo video for this submission, so I'll document that gap in the PR. Code + tests are the deliverable.

**Comment by phquand2000** [comment]
__acb_probe__

**Comment by akamabu** [comment]
/attempt #59

I will implement an R2-backed asynchronous export job path for large database dumps, with resumable chunked processing and tests, while preserving synchronous behavior where appropriate.

**Comment by akamabu** [comment]
I have submitted a PR for this issue: https://github.com/outerbase/starbasedb/pull/202

The implementation:
- Replaces in-memory dump with a `ReadableStream` using `LIMIT/OFFSET` pagination (no full-DB load)
- Adds `?async=true` mode that writes to R2 and returns `202 Accepted` immediately, using `ctx.waitUntil` so the upload continues after the response
- Supports custom filename, callback URL notification on completion, and configurable chunk size
- Properly escapes SQL identifiers and values (NULL, booleans, strings, blobs)

**Comment by akamabu** [comment]
Opened PR: https://github.com/outerbase/starbasedb/pull/202

**Comment by Treasure520520** [comment]
/attempt #59

I am taking a focused SQL dump streaming slice: replace the current full in-memory SQL dump string with a paged ReadableStream response for `/export/dump`, preserving existing response behavior while avoiding one giant database-sized string in Durable Object memory. I will keep this narrow and tested so it can complement larger async/R2 submissions.

**Comment by BobbyZ2026** [comment]
/attempt #59

Implementing a focused streaming export refactor for the current export endpoints:

- Replace full Blob/string buffering for SQL, CSV, and JSON exports with ReadableStream responses.
- Page table reads so only one bounded batch is materialized at a time.
- Prefer keyset pagination on primary keys/rowid, with OFFSET fallback only when cursor values are unavailable.
- Quote SQLite identifiers, serialize SQL values safely, preserve CSV escaping, and add regression tests for the streaming helpers and routes.

**Comment by Mira-Mjodheim** [comment]
Hi @starbasedb team, I'd be happy to help investigate the issue with database dumps on large databases. Can you provide more details about the problem you're experiencing, such as error messages or the size of the databases that are affected?

**Comment by quangnd2203** [comment]
Hi! I noticed there is already a linked PR and an Algora attempt on this bounty. Is this still open for a new contributor, or should it be considered taken for now? Thanks!

**Comment by Mira-Mjodheim** [comment]
Hi, I'd be happy to help investigate the issue with database dumps on large databases. Can you provide more details about the error you're encountering or the expected behavior vs. actual behavior?

**Comment by swhan0329** [comment]
/attempt #59

Plan: submit a narrow SQL dump correctness fix for table names that require escaping. The current dump path interpolates table names directly into sqlite_master lookups, SELECTs, and generated INSERT statements, which breaks dumps for tables with spaces, quotes, or SQLite keywords. I will add small identifier/string escaping helpers plus focused regression coverage, keeping the change isolated to src/export/dump.ts and its test.

**Comment by swhan0329** [comment]
/attempt #59

Second scoped follow-up: CSV/JSON exports share getTableData(), which currently checks table existence with a parameterized sqlite_master query but then runs SELECT using the raw table name. I am submitting a separate small fix so table names with spaces, quotes, or SQLite keywords export correctly through the CSV/JSON routes as well, without changing the broader streaming/R2 export work.

**Comment by kartikganesh** [comment]
/attempt #59

**Implementation plan:**
- Add R2 binding for backup storage
- Chunked export processing (1000 rows/batch) to avoid memory pressure
- DO alarm-based continuation for exports exceeding 25s
- Progress tracking via DO storage with resume capability
- Status endpoint (`/export/status/:id`) and download endpoint (`/export/download/:id`)
- Callback URL notification on completion
- Fallback to current in-memory behavior when R2 is not configured

**Comment by harishn1408** [comment]
I have a complete implementation ready for this issue.

**Full patch available as a Gist (ready to apply):**
https://gist.github.com/harishn1408/85354eeb588994a1f15fad2d128c2a26

## What the patch does

**src/export/dump.ts** — Two-pronged approach:

1. **GET /export/dump — ReadableStream** (O(1) memory)
   - Pages rows 200 at a time via SQLite cursor
   - Streams SQL header → per-table CREATE+INSERTs → footer without ever holding the full dump in memory
   - Safe value escaping (escapeSqlValue) for all types: null, bigint, boolean, string

2. **POST /export/dump/r2 — Async R2 multipart upload**
   - Returns 202 Accepted immediately; background work via ctx.waitUntil
   - Chunked S3-compatible multipart upload to DUMP_BUCKET R2 binding
   - Generates a signed download URL returned in the JSON response

**src/handler.ts** — Registers the new /export/dump/r2 POST route.

## Why this approach
- Zero full-DB memory allocation at any point
- Works within Cloudflare Workers' 30s CPU limit via background task
- Backward compatible — existing GET /export/dump still works

My GitHub account has a temporary fork restriction, so I cannot open a PR directly. The patch can be applied with:
`ash
curl -s https://gist.githubusercontent.com/harishn1408/85354eeb588994a1f15fad2d128c2a26/raw/streaming-dump-fix.patch | git apply
`

Happy to answer any questions or adjust the implementation.

**Comment by ThaiTrevor** [comment]
C:/Program Files/Git/attempt #59

**Comment by kartikganesh** [comment]
@Brayden — just checking in. I submitted PR #238 with full R2 multipart streaming, DO alarm continuation, and 10 unit tests. Happy to address any feedback. Let me know if this is still being evaluated.

**Comment by PTHAICAP** [comment]
/attempt #59

Hi team! I've already opened a PR for this bounty: 
https://github.com/outerbase/starbasedb/pull/249

It implements streaming + pagination for /export/dump as 
requested. Ready for review. Thanks!

**Comment by PTHAICAP** [comment]
Hi! Submitting a PR for review related to this issue — handles 
large database dumps via streaming + pagination. Feedback welcome!

PR link: https://github.com/outerbase/starbasedb/pull/249

**Comment by chaudl113** [comment]
/attempt #59

**Plan:**
- Stream-based export that doesn't load entire DB into memory
- Chunked processing with configurable batch size
- Support for SQL, CSV, and JSON formats
- R2 integration for large dumps that exceed response timeout
- Breathing intervals to avoid DO lock contention

Will start with a streaming SQL export implementation.

**Comment by chaudl113** [comment]
Submitted PR #255 with streaming implementation:

- **Streaming via TransformStream** — no more loading entire DB into memory
- **Chunked LIMIT/OFFSET queries** — 1000 rows per batch
- **Breathing intervals** — 10ms between chunks to avoid DO lock contention
- **Works for SQL, CSV, and JSON exports**
- **All 23 export tests pass** including new chunked streaming test

The key fix: instead of  (loads everything), we now do  in a loop, writing each batch directly to the response stream.

**Comment by chaudl113** [comment]
/attempt #59

Plan:
- Implement stream-based export using chunked processing with breathing intervals
- Use Cloudflare R2 for temporary storage of large dumps
- Deliver via signed URL download
- Support SQL, CSV, JSON formats
- Prevent DO lockup with configurable work/pause cycles

**Comment by chaudl113** [comment]
/claim #59

PR submitted: https://github.com/outerbase/starbasedb/pull/259

Stream-based export with TransformStream, paginated queries (500 rows/batch), and breathing intervals. All tests pass.

**Comment by watcharaponthod-code** [comment]
/attempt #59

I've opened PR #260 which fixes the large-database export timeout by replacing the in-memory dump approach with a TransformStream-based streaming implementation that fetches rows in LIMIT/OFFSET chunks of 1000. See https://github.com/outerbase/starbasedb/pull/260 for full details.

**Comment by FishEnjoyer2025** [comment]
/attempt #59

Streaming the dump via a ReadableStream and reading each table in bounded LIMIT/OFFSET batches, so large databases dump with roughly constant memory and without hitting the 30s request window. PR incoming.

**Comment by Vinzz2303** [comment]
I've submitted a PR for this issue: https://github.com/outerbase/starbasedb/pull/270. The fix uses ReadableStream with LIMIT/OFFSET batching to handle large databases without memory exhaustion or gateway timeouts. All 9 tests pass.

**Comment by OP-88** [comment]
/attempt #59

Implementation Plan:

- Detect DB size on export request; if small enough to complete under 30s, return directly as before

- For large exports, initiate a chunked background export using DO alarms to resume across timeout boundaries

- Write chunks progressively to R2 with filename format dump_YYYYMMDD-HHmmss.sql

- Track cursor position in DO internal storage so alarms can resume from where they left off

- Add breathing intervals to avoid locking the DO for other requests

- On completion, hit a user-provided callback URL with the R2 file location

- Expose a /export/status/:jobId endpoint for polling

**Comment by Shaidyk** [comment]
/claim #59

PR: https://github.com/outerbase/starbasedb/pull/279

**Comment by hiSandog** [comment]
The breathing-interval idea probably needs an explicit export job state machine. For large dumps, it would help to track cursor/page token, output byte offset, format-specific footer state, and cancellation/error state separately from the HTTP request that starts the export. That would let SQL/CSV/JSON share the same resumable streaming core while still handling each format's header/footer and escaping rules correctly.

**Comment by Furox-Art** [comment]
C:/Program Files/Git/attempt #59

**Comment by Furox-Art** [comment]
Full implementation is up: #298 🚀

- Chunked, resumable dump engine (`ChunkedDumpEngine`): rowid-batched scans, time-budgeted cycles with breathing intervals, crash-safe progress in DO storage
- R2 spill when a bucket is bound + streamed reassembly (client never holds the full dump in memory)
- >30s dumps upgrade to a DO-alarm-driven job; `GET /export/dump/status` for progress + download
- Small dumps keep the existing synchronous response (backwards compatible)
- **12 unit tests passing** (`chunkedDump.test.ts`), including resume-mid-schema without duplicate emission and SQL-injection-safe identifier filtering
